iree/compiler/Conversion/Common/LinalgBufferizePass.cpp

```diff
@@ -49,458 +50,528 @@ namespace mlir {
-  Value newInputBuffer = bvm.lookup(op.source());
-  if (auto tensorType =
-          op->getResult(0).getType().dyn_cast<RankedTensorType>()) {
-    // If the op return a Tensor allocate a buffer for the returned value.
-    auto tensorShape = tensorType.getShape();
-    SmallVector<Value, 4> dynOperands;
-    for (size_t idx : llvm::seq(size_t(0), tensorShape.size())) {
-      if (tensorType.isDynamicDim(idx)) {
-        Value tensor = bvm.lookupOrNull(outputTensor);
-        if (!tensor) tensor = outputTensor;
-        dynOperands.push_back(b.createOrFold<memref::DimOp>(loc, tensor, idx));
+  RankedTensorType resultType =
+      op->getResult(0).getType().cast<RankedTensorType>();
+  SmallVector<Value, 4> dynamicDims;
+
+  // Get the shape of the result
+  Location loc = op->getLoc();
+  if (auto shapedOp = dyn_cast<InferShapedTypeOpInterface>(op)) {
+    SmallVector<SmallVector<Value>> resultShape;
+    if (failed(shapedOp.reifyReturnTypeShapesPerResultDim(b, resultShape))) {
+      return nullptr;
+    }
+    for (auto shape : enumerate(resultShape[0])) {
+      if (resultType.isDynamicDim(shape.index())) {
+        dynamicDims.push_back(shape.value());
       }
     }
-    auto alloc = allocationFn(b, loc, tensorShape, tensorType.getElementType(),
-                              dynOperands);
-    bvm.map(op->getResult(0), alloc);
-    transferShapeOpsToMemref(b, op->getResult(0), alloc, bvm);
-  }
-
-  // Replace the tensor operand.
-  if (auto readOp = dyn_cast<vector::TransferReadOp>(op.getOperation())) {
-    readOp.sourceMutable().assign(newInputBuffer);
+  } else if (auto subTensorOp = dyn_cast<SubTensorOp>(op)) {
+    dynamicDims = llvm::to_vector<4>(subTensorOp.sizes());
+  } else if (auto subTensorInsertOp = dyn_cast<SubTensorInsertOp>(op)) {
+    dynamicDims = getDynamicDims(b, loc, subTensorInsertOp.dest());
+  } else if (auto transferWriteOp = dyn_cast<vector::TransferWriteOp>(op)) {
+    dynamicDims = getDynamicDims(b, loc, transferWriteOp.source());
   } else {
-    auto writeOp = cast<vector::TransferWriteOp>(op.getOperation());
-    // Create a new transfer_write on buffer that doesn't have a return value.
-    // Leave the previous transfer_write to dead code as it still has uses at
-    // this point.
-    b.create<vector::TransferWriteOp>(
-        loc, writeOp.vector(), newInputBuffer, writeOp.indices(),
-        writeOp.permutation_map(), writeOp.mask(),
-        writeOp.in_bounds() ? *writeOp.in_bounds() : ArrayAttr());
+    return nullptr;
   }
-  return success();
+  return allocationFn(b, loc, resultType.getShape(),
+                      resultType.getElementType(), dynamicDims);
 }
 
-// Extract int64_t values from the assumed ArrayAttr of IntegerAttr.
-static SmallVector<int64_t, 4> extractFromI64ArrayAttr(Attribute attr) {
-  return llvm::to_vector<4>(llvm::map_range(
-      attr.cast<ArrayAttr>(),
-      [](Attribute a) -> int64_t { return a.cast<IntegerAttr>().getInt(); }));
+template <typename TensorType>
```

Comment:
can it be anything else than RankedTensorType?

Reply:
Could be `flow.dispatch.tensor` as well.